File: ui/visual-component/src/main/java/org/visual/component/control/UndecoratedDialog.java
package org.visual.component.control;

import javafx.scene.control.Dialog;
import javafx.scene.input.MouseEvent;
import javafx.stage.StageStyle;

public class UndecoratedDialog<R> extends Dialog<R> {
  private double xOffset = 0.0;
  private double yOffset = 0.0;

  public UndecoratedDialog() {
    initStyle(StageStyle.TRANSPARENT);
    setResizable(true);

    dialogPaneProperty()
        .addListener(
            (observable, oldValue, dialogPane) -> {
              dialogPane.setOnMousePressed(this::onMousePressed);
              dialogPane.setOnMouseDragged(this::onMouseDragged);
              dialogPane.setOnMouseReleased(this::onMouseReleased);
              dialogPane.setStyle("-fx-border-color: black; -fx-border-width: 1;");
            });
  }

  private void onMousePressed(MouseEvent event) {
    xOffset = event.getSceneX();
    yOffset = event.getSceneY();
  }

  private void onMouseDragged(MouseEvent event) {
    setX(event.getScreenX() - xOffset);
    setY(event.getScreenY() - yOffset);
  }

  private void onMouseReleased(MouseEvent event) {
    getDialogPane().setOpacity(1.0);
  }
}

File: ui/visual-component/src/main/kotlin/org/visual/component/control/UndecoratedDialog.kt
package org.visual.component.control

import javafx.scene.control.Dialog
import javafx.scene.input.MouseEvent
import javafx.stage.StageStyle

class UndecoratedDialog<R> : Dialog<R>() {

  private var xOffset = 0.0
  private var yOffset = 0.0

  init {
    initStyle(StageStyle.TRANSPARENT)
    isResizable = true
    dialogPaneProperty().addListener { _, _, _ ->
      run {
        dialogPane.setOnMousePressed(this::onMousePressed)
        dialogPane.setOnMouseDragged(this::onMouseDragged)
        dialogPane.setOnMouseReleased(this::onMouseReleased)
      }
    }
  }

  private fun onMousePressed(event: MouseEvent) {
    xOffset = event.sceneX
    yOffset = event.sceneY
  }

  private fun onMouseDragged(event: MouseEvent) {
    x = event.screenX - xOffset
    y = event.screenY - yOffset
  }

  private fun onMouseReleased(event: MouseEvent) {
    dialogPane.opacity = 1.0
  }
}
